src/group_theory/torsion.lean

```diff
@@ -75,3 +77,49 @@ exponent_exists_iff_ne_zero.mpr $
 @[to_additive is_add_torsion_of_fintype]
 lemma is_torsion_of_fintype [fintype G] : is_torsion G :=
 exponent_exists.is_torsion $ exponent_exists_iff_ne_zero.mpr exponent_ne_zero_of_fintype
+
+section infinite_torsion_groups
+
+/-- The group ℚ ⧸ ℤ. -/
+@[reducible] def qmodz := ℚ ⧸ (algebra.linear_map ℤ ℚ).range
+
+namespace qmodz
+
+open rat
+
+/-- ℚ ⧸ ℤ is a torsion group -/
+lemma is_torsion : add_monoid.is_torsion qmodz :=
+λ qz, quotient.induction_on' qz $ begin
+  intro q,
+  rw is_of_fin_add_order_iff_nsmul_eq_zero,
+  refine ⟨q.denom, q.pos, quotient.eq'.mpr _⟩,
+  by_cases h : q.denom = 0,
+  { rw h, simp only [nsmul_eq_mul, nat.cast_zero, zero_mul], exact quotient.eq'.mp rfl },
+  { use q.num,
+    apply @mul_right_cancel₀ _ _ _ (q.denom : ℚ)⁻¹ _
+      (by simp only [h, ne.def, inv_eq_zero, not_false_iff, nat.cast_eq_zero]),
+    have z' : ↑(q.num) * (↑(q.denom)) ⁻¹ = (q.num : ℚ) / q.denom := by rw div_eq_mul_inv,
+    field_simp [z', mul_comm, ← _root_.mul_assoc], }
+end
+
+end qmodz
+
+namespace polynomial
+
+open_locale polynomial
+
+variables {Fq : Type*} [field Fq] [fintype Fq]
+
+/-- Fq[X] is a torsion group -/
+lemma is_torsion : add_monoid.is_torsion Fq[X] := begin
+  intro f,
+  rw is_of_fin_add_order_iff_nsmul_eq_zero,
+  refine ⟨fintype.card Fq, ⟨fintype.card_pos, _⟩⟩,
+  by_cases h : f = 0, by rw [h, smul_zero],
+  { rw [nsmul_eq_mul, mul_eq_zero], left,
+    rw [← C_eq_zero.mpr (finite_field.cast_card_eq_zero Fq), C_eq_nat_cast] },
+end
```

Comment:
```suggestion
lemma is_torsion : add_monoid.is_torsion Fq[X] :=
λ f, begin
  rw is_of_fin_add_order_iff_nsmul_eq_zero,
  refine ⟨fintype.card Fq, fintype.card_pos, _⟩,
  rw [nsmul_eq_smul_cast Fq _ f, finite_field.cast_card_eq_zero, zero_smul],
end
```
There's no point doing `by_cases h : foo` if one of your cases doesn't ever use `h`! In this case, `nsmul_eq_smul_cast` does all the work for you.

Reply:
Maybe I just need to think harder to see why that's the case (thanks of course for th

Reply:
> Maybe I just need to think harder to see why that's the case

Your proof was "either we use the fact that `f = 0`, or we provide a solution that works in all cases (including `f = 0`)" - so just skip to the second step!

Reply:
I see, right, thank you!

Reply:
Your other proof had a different failing in `by_cases`: `q.denom = 0` is impossible, according to `rat.denom_ne_zero`, so casing on it is like `by_cases : false`.